azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationService.kt

```diff
@@ -0,0 +1,272 @@
+package com.anaplan.engineering.azuki.script.generation
+
+import com.anaplan.engineering.azuki.core.system.*
+import com.anaplan.engineering.azuki.declaration.*
+
+/**
+ * The main endpoint for script generation tasks.
+ */
+class ScriptGenerationService<
+    // mandatory factories
+    out AF : ActionFactory,
+    out CF : CheckFactory,
+    // optional factories (these require calling 'withXYZFactory')
+    out QF : QueryFactory,
+    out AGF : ActionGeneratorFactory,
+    // mandatory non-factory parameters
+    DS : DeclarationState,
+    // optional non-factory parameters (these require calling 'withXYZ')
+    E : ScriptGenerationEnvironment,
+    > private constructor(
+    // mandatory parameters
+    internal val actionFactory: AF,
+    internal val checkFactory: CF,
+    internal val declarationStateFactory: DeclarationStateFactory<DS>,
+    // optional parameters
+    internal val environmentFactory: ScriptGenerationEnvironmentFactory<E>,
+    internal val queryQueryFactory: QF,
+    internal val verifyQueryFactory: QF,
+    internal val actionGeneratorFactory: AGF,
+) {
+
+    /**
+     * Provides ready-made usage patterns for script generation of a whole scenario, from a whole scenario.
+     */
+    val scenario = ScriptGenerationScenarioPatterns(this)
```

Comment:
❓ This feels like a layering problem: `ScriptGenerationScenarioPatterns` is a higher-level abstraction that we're depending on in this lower-level construct.

Previously, I had all of its constituent methods written as extension methods on `ScriptGenerationService`.  That felt a bit more honest, but it also seemed inelegant to introduce a load of extensions purely for layering purposes...

Can I get a second opinion on this?